File: challenge/leet/easy/q118.py
'''
q 118 Pascal's Triangle
https://leetcode.com/problems/pascals-triangle/description/
'''
import unittest
from typing import List


class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        tri = [[1]]
        for r in range(1, numRows):
            c0 = len(tri[-1])
            row = [1] * (c0 + 1)
            for c in range(1, c0):
                row[c] = tri[-1][c-1] + tri[-1][c]
            tri.append(row)
        return tri;


class MyTestCase(unittest.TestCase):
    def test_something(self):
        s = Solution()
        self.assertEqual([[1]], s.generate(1))
        self.assertEqual([[1],[1,1]], s.generate(2))
        self.assertEqual([[1],[1,1],[1,2,1]], s.generate(3))


if __name__ == '__main__':
    unittest.main()

File: challenge/leet.gcc/easy/q118/q118.cpp
/**
 * 118. Pascal's Triangle
 * https://leetcode.com/problems/pascals-triangle/
 */

#include <vector>
#include <iostream>
#include <string>
#include <numeric>

using namespace std;

class Solution {
public:
    vector<vector<int>> getRow(int numRows) {
        vector<vector<int>> triangle{ vector<int>{1} };
        for (int r = 1; r < numRows; r++) {
            vector<int> row{ 1 };
            for (int c = 1; c < triangle[triangle.size() - 1].size(); c++) {
                row.push_back(triangle[r-1][c-1] + triangle[r-1][c]);
            }
            row.push_back(1);
            triangle.push_back(row);
        }
        return triangle;
    }
};

string reduce(vector<int> q) {
    return reduce(next(q.begin()), q.end(), string("[" + to_string(q[0])), [](string a, int b) { return a + ", " + to_string(b); }) + string("]");
}

int main() {
    Solution s;

    vector<vector<int>> v;

    v = s.getRow(1);
    cout << "1 ? " << endl << reduce(next(v.begin()), v.end(), string(reduce(v[0])), [](string p, vector<int>& q) { return move(p) + ",\n" + reduce(q); }) << endl;

    v = s.getRow(2);
    cout << "2 ? " << endl << reduce(next(v.begin()), v.end(), string(reduce(v[0])), [](string p, vector<int>& q) { return move(p) + ",\n" + reduce(q); }) << endl;

    v = s.getRow(3);
    cout << "3 ? " << endl << reduce(next(v.begin()), v.end(), string(reduce(v[0])), [](string p, vector<int>& q) { return move(p) + ",\n" + reduce(q); }) << endl;

    v = s.getRow(4);
    cout << "3 ? " << endl << reduce(next(v.begin()), v.end(), string(reduce(v[0])), [](string p, vector<int>& q) { return move(p) + ",\n" + reduce(q); }) << endl;

    v = s.getRow(5);
    cout << "3 ? " << endl << reduce(next(v.begin()), v.end(), string(reduce(v[0])), [](string p, vector<int>& q) { return move(p) + ",\n" + reduce(q); }) << endl;

    return 0;
}
